### DISCUSSION TITLE
redundant parenthesis

### DISCUSSION THREAD

**Discussion by ningshuofu** [sqlalchemy/sqlalchemy#10323] [Usage Questions] [answered]
```
n = model.id.not_in(['1'])
s = select(model.in).where(n)
db.session.execute(s)
db.session.commit()
```
row sql:
```
SELECT
  model.id
FROM
  model
WHERE
  (model.id NOT IN ('1'))
```
i need:
```
SELECT
  model.id
FROM
  model
WHERE
  model.id NOT IN ('1')
```
ps: `in` has not problem

**Comment by CaselIT** [comment]
Hi,

the two sql should be equivalent. What is the issue?

**Reply by ningshuofu** [reply]
yes, the two sql is equivalent, but 
```
SELECT
  model.id
FROM
  model
WHERE
  (model.id NOT IN ('1'))
```
should not has redundant parenthesis, i want to know how to remove it and why

**Reply by ningshuofu** [reply]
@CaselIT

**Reply by CaselIT** [reply]
there isn't a way of changing it currently.

**Reply by ningshuofu** [reply]
thx anyway

**Reply by ningshuofu** [reply]
but why in does not add brackets and not_in adds brackets is still very confusing.

**Comment by zzzeek** [comment]
@CaselIT hmm why is it doing that?  it doesnt do it for IN

**Reply by zzzeek** [reply]
oh because it's a not() of the whole thing

**Reply by ningshuofu** [reply]
i try use ~model.id.in(['1']), not_(model.id.in(['1'])) and more, it is no use

**Reply by zzzeek** [reply]
oh OK it's hardcoded in the compiler to suit the empty IN case, OK.

**Reply by ningshuofu** [reply]
row sql should not has '()', can you explain more, thx

**Reply by zzzeek** [reply]
why not ?  it's valid for any SQL parser

**Comment by zzzeek** [comment, marked as answer]
yeah, OK this compilation rule is specific to NOT IN and is an implementation detail of how SQLAlchemy would compile the case for "NOT IN" using an empty list, so it can't be changed right now.

this would only be an issue if you are doing some kind of programmatic SQL parsing, so just modify your parser to accommodate these parens for the NOT IN case.

**Reply by ningshuofu** [reply]
thx, Can you give me some source code or link to prove that not in must add brackets?

**Reply by CaselIT** [reply]
oh we added that when reworking the empty in since that needs an `OR` condition.
I forgot about it

**Reply by zzzeek** [reply]
oh right, it's for like PostgreSQL where we are doing that silly parenthesis trick:

```
SELECT a.id 
FROM a 
WHERE (a.id NOT IN (NULL) OR (1 = 1))
```

**Reply by CaselIT** [reply]
I think that we do that for most db

**Comment by zzzeek** [comment]
> thx, Can you give me some source code or link to prove that not in must add brackets?

Here's a test; run this on PostgreSQL or MySQL/MariaDB, not with SQLite:

```py
from __future__ import annotations

from sqlalchemy import create_engine, select, String
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(50))


#e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
e = create_engine("mysql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)


with Session(e) as session:
    session.add(A(name="jack"))
    session.flush()

    result = session.execute(
        select(A.id).where(A.name == "ed").where(A.id.not_in([]))
    )
    assert not result.first()

```

run the program, succeeds.

apply patch:

```diff
diff --git a/lib/sqlalchemy/sql/compiler.py b/lib/sqlalchemy/sql/compiler.py
index 314cbe2167..769d28df3d 100644
--- a/lib/sqlalchemy/sql/compiler.py
+++ b/lib/sqlalchemy/sql/compiler.py
@@ -3095,7 +3095,7 @@ class SQLCompiler(Compiled):
         # The brackets are required in the NOT IN operation because the empty
         # case is handled using the form "(col NOT IN (null) OR 1 = 1)".
         # The presence of the OR makes the brackets required.
-        return "(%s)" % self._generate_generic_binary(
+        return "%s" % self._generate_generic_binary(
             binary, OPERATORS[operator], **kw
         )
 

```

run program again, it fails:

```
Traceback (most recent call last):
  File "/mnt/photon_home/classic/dev/sqlalchemy/test3.py", line 34, in <module>
    assert not result.first()
AssertionError

```

**Reply by ningshuofu** [reply]
thank your help! you are a serious person.
